design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,173 @@
+typically deployed together. It is unlikely to encounter a Swift
+object store API outside an OpenStack cloud and in the context of an
+OpenStack cloud the Swift API needs to be integrated with the Keystone
+authentication service.
+
+Any changes to the CRD must be future safe and cleanly allow extension
+to further technologies (such as LDAP authentication).
+
+### Non-Goals
+
+* Support for Keystone API versions below v3. API version v2 has long
+  been deprecated [has been removed in
+  Queens](https://docs.openstack.org/keystone/latest/contributor/http-api.html)
+  which was release in 2018 and is now in extended maintenance mode
+  (which means it gets no more points releases and only sporadic bug
+  fixes/security fixes).
+
+* Authenticating Ceph RGW to Keystone via admin token – Only
+  authentication via an OpenStack service account will be supported.
+
+* Support for APIs beside S3 and Swift.
+
+* Support for authentication technologies other than Keystone (e.g. LDAP)
+
+* Exposing options that disable security features (e.g. TLS verification)
+
+## Proposal details
+
+The Object Store CRD will have to be extended to accommodate the new
+settings.
+
+### Keystone integration
+
+A new section `auth:` is added to the Object Store CRD. To configure
+the keystone integration:
+
+```yaml
+auth:
+  keystone:
+    url: https://keystone:5000/
+    acceptedRoles: ["_member_", "service", "admin"]
+    apiVersion: 3
+    implicitTenants: swift
+    tokenCacheSize: 1000
+    revocationInterval: 1200
+    serviceUserSecret: rgw-service-user
+```
+
+The credentials for the Keystone service account used by Ceph RGW are
+supplied in a Secret that contains a mapping of OpenStack openrc
+environment variables. Only password authentication to Keystone is
+supported. Example:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: rgw-service-user
+data:
+  OS_PASSWORD: "horse staples battery correct"
+  OS_USERNAME: "ceph-rgw"
+  OS_PROJECT_NAME: "admin"
+  OS_USER_DOMAIN_NAME: "Default"
+  OS_PROJECT_DOMAIN_NAME: "Default"
+  OS_AUTH_TYPE: "password"
+```
+
+This format is chosen because it is a natural and interoperable way
+that keystone credentials are represented and for the supported auth
+type it maps naturally to the Ceph RGW configuration.
+
+The following constraints must be fulfilled by the secret:
+* `OS_AUTH_TYPE` must be `password` or omitted.
+* `OS_USER_DOMAIN_NAME` must equal `OS_PROJECT_DOMAIN_NAME`.
+* All other openrc variables (e.g. API version and endpoint) are ignored.
+
+### Swift integration
+
+Swift is configured in the `gateway:` section of the Object Store CRD.
+
+The currently unused `type:` argument may now take the values `swift`
+or `s3`.
+
+The following new settings are available to properly configure Swift:
+```yaml
+gateway:
+  type: swift
+  swiftAccountInUrl: true
+  swiftUrlPrefix: /swifter
+  swiftVersioningEnabled: false
+```
+
+### Risks and Mitigation
+
```

Comment:
There is cephobjectstoreuser CRD which represents s3 user, it automatically generates a secret containing information about accessing the RGW server via s3. Are you planning something similar inline, Either integrate with this user or create New CR say `cephobjectsoreswiftuser`.  Also what about plan with OB/OBCs ?

Reply:
Thanks for the pointers, I missed that point because we don't have Swift users in our use-case that are managed by Keystone. I've addressed it in the new commits.